Fix warning: include/wx/textbuf.h(45): warning C4191: 'type cast': un…

rbresalier opened the pull request:
…safe conversion from 'wxArrayLinesType::CMPFUNC' to 'wxArrayLinesType::SCMPFUNC'.

When using wxUSE_STD_CONTAINERS as 1, and utilizing wxXmlDocument, the wxXmlDocument utilizes textbuf.h, and the warning comes out. I did not have this warning with my code using wxWidgets 3.1.0, but started getting it when I switched to wxWidgets 3.1.1. The warning started happening for me due to commit:

https://github.com/wxWidgets/wxWidgets/commit/11e54135587d2073c4f1c0986e072078678bf229

Which introduced the usage of textbuf.h into xml.h. However the underlying problem has been present long before this commit.

Due to heavy preprocessor use, the true source of where the warning is coming from is hard to track. I tracked the warning down to line 170 in include/wx/dynarray.h:

170: Predicate p((SCMPFUNC)fnCompare);

SCMPFUNC and fnCompare (of type CMPFUNC) have different interfaces.

SCMPFUNC has interface:
112: int ()(T pItem1, T pItem2)

while CMPFUNC has interface:
89: int()(T* pItem1, T* pItem2)

So SCMPFUNC is taking pItem1 and pItem2 not as pointers, but as value, while CMPFUNC takes them as pointers.

As CMPFUNC is what is used in the API to clients, this is the API we want to preserve. The SCMPFUNC API is not useful.

To fix the issue, I change the definition of Predicate to be a functor that has the CMPFUNC interface, which is to change lines 112-113 from:

   typedef int (wxCMPFUNC_CONV *CMPFUN##name)(T pItem1, T pItem2);  \
   typedef wxSortedArray_SortFunction<T, CMPFUN##name> name##_Predicate; \

to:

   typedef wxArray_SortFunction<T> name##_Predicate; \

Now the code compiles without warning and is correct.

Backup:

To track down the interfaces of CMPFUNC and SCMPFUNC, here is a guide:

CMPFUNC:
122: typedef wxArray_SortFunction<T>::CMPFUNC CMPFUNC;
92: bool operator()(const T& i1, const T& i2)
89: int (wxCMPFUNC_CONV *CMPFUNC)(T* pItem1, T* pItem2)

SCMPFUNC:
120: typedef predicate::CMPFUNC SCMPFUNC;
119: typedef predicate Predicate;
113: typedef wxSortedArray_SortFunction<T, CMPFUN##name> name##_Predicate;
112: typedef int (wxCMPFUNC_CONV *CMPFUN##name)(T pItem1, T pItem2)

Commit:
Fix warning: include/wx/textbuf.h(45): warning C4191: 'type cast': unsafe conversion from 'wxArrayLinesType::CMPFUNC' to 'wxArrayLinesType::SCMPFUNC'.

When using wxUSE_STD_CONTAINERS as 1, and utilizing wxXmlDocument, the wxXmlDocument utilizes textbuf.h, and the warning comes out. I did not have this warning with my code using wxWidgets 3.1.0, but started getting it when I switched to wxWidgets 3.1.1. The warning started happening for me due to commit:

https://github.com/wxWidgets/wxWidgets/commit/11e54135587d2073c4f1c0986e072078678bf229

Which introduced the usage of textbuf.h into xml.h. However the underlying problem has been present long before this commit.

Due to heavy preprocessor use, the true source of where the warning is coming from is hard to track. I tracked the warning down to line 170 in include/wx/dynarray.h:

170: Predicate p((SCMPFUNC)fnCompare);

SCMPFUNC and fnCompare (of type CMPFUNC) have different interfaces.

SCMPFUNC has interface:
112: int ()(T pItem1, T pItem2)

while CMPFUNC has interface:
89: int()(T* pItem1, T* pItem2)

So SCMPFUNC is taking pItem1 and pItem2 not as pointers, but as value, while CMPFUNC takes them as pointers.

As CMPFUNC is what is used in the API to clients, this is the API we want to preserve. The SCMPFUNC API is not useful.

To fix the issue, I change the definition of Predicate to be a functor that has the CMPFUNC interface, which is to change lines 112-113 from:

   typedef int (wxCMPFUNC_CONV *CMPFUN##name)(T pItem1, T pItem2);  \
   typedef wxSortedArray_SortFunction<T, CMPFUN##name> name##_Predicate; \

to:

   typedef wxArray_SortFunction<T> name##_Predicate; \

Now the code compiles without warning and is correct.

Backup:

To track down the interfaces of CMPFUNC and SCMPFUNC, here is a guide:

CMPFUNC:
122: typedef wxArray_SortFunction<T>::CMPFUNC CMPFUNC;
92: bool operator()(const T& i1, const T& i2)
89: int (wxCMPFUNC_CONV *CMPFUNC)(T* pItem1, T* pItem2)

SCMPFUNC:
120: typedef predicate::CMPFUNC SCMPFUNC;
119: typedef predicate Predicate;
113: typedef wxSortedArray_SortFunction<T, CMPFUN##name> name##_Predicate;
112: typedef int (wxCMPFUNC_CONV *CMPFUN##name)(T pItem1, T pItem2)

Commit:
Fix warning: include/wx/textbuf.h(45): warning C4191: 'type cast': unsafe conversion from 'wxArrayLinesType::CMPFUNC' to 'wxArrayLinesType::SCMPFUNC'.

When using wxUSE_STD_CONTAINERS as 1, and utilizing wxXmlDocument, the wxXmlDocument utilizes textbuf.h, and the warning comes out. I did not have this warning with my code using wxWidgets 3.1.0, but started getting it when I switched to wxWidgets 3.1.1. The warning started happening for me due to commit:

https://github.com/wxWidgets/wxWidgets/commit/11e54135587d2073c4f1c0986e072078678bf229

Which introduced the usage of textbuf.h into xml.h. However the underlying problem has been present long before this commit.

Due to heavy preprocessor use, the true source of where the warning is coming from is hard to track. I tracked the warning down to line 170 in include/wx/dynarray.h:

170: Predicate p((SCMPFUNC)fnCompare);

SCMPFUNC and fnCompare (of type CMPFUNC) have different interfaces.

SCMPFUNC has interface:
112: int ()(T pItem1, T pItem2)

while CMPFUNC has interface:
89: int()(T* pItem1, T* pItem2)

So SCMPFUNC is taking pItem1 and pItem2 not as pointers, but as value, while CMPFUNC takes them as pointers.

As CMPFUNC is what is used in the API to clients, this is the API we want to preserve. The SCMPFUNC API is not useful.

To fix the issue, I change the definition of Predicate to be a functor that has the CMPFUNC interface, which is to change lines 112-113 from:

   typedef int (wxCMPFUNC_CONV *CMPFUN##name)(T pItem1, T pItem2);  \
   typedef wxSortedArray_SortFunction<T, CMPFUN##name> name##_Predicate; \

to:

   typedef wxArray_SortFunction<T> name##_Predicate; \

Now the code compiles without warning and is correct.

Backup:

To track down the interfaces of CMPFUNC and SCMPFUNC, here is a guide:

CMPFUNC:
122: typedef wxArray_SortFunction<T>::CMPFUNC CMPFUNC;
92: bool operator()(const T& i1, const T& i2)
89: int (wxCMPFUNC_CONV *CMPFUNC)(T* pItem1, T* pItem2)

SCMPFUNC:
120: typedef predicate::CMPFUNC SCMPFUNC;
119: typedef predicate Predicate;
113: typedef wxSortedArray_SortFunction<T, CMPFUN##name> name##_Predicate;
112: typedef int (wxCMPFUNC_CONV *CMPFUN##name)(T pItem1, T pItem2)

Commit:
Remove wxSortedArray_SortFunction as it is dead code.

It is no longer needed after commit https://github.com/rbresalier/wxWidgets/commit/f15387f9202e2d66406facff4a106fe0ffa33a30

Commit:
Remove wxSortedArray_SortFunction as it is dead code.

It is no longer needed after commit https://github.com/rbresalier/wxWidgets/commit/f15387f9202e2d66406facff4a106fe0ffa33a30

Commit:
Remove _WX_DECLARE_BASEARRAY_2 as it is not useful and can cause problems.

Commit:
Remove _WX_DECLARE_BASEARRAY_2 as it is not useful and can cause problems.

PBfordev committed:
Clarify wxSL_SELRANGE description

The user cannot select a range, the selection range can be changed only programatically.

PBfordev committed:
Fix memory leak in taskbarbutton sample

wxTaskBarButton::RemoveThumbBarButton() removes the button but does not
delete it. In the sample removed buttons must be deleted manually
because they are also removed from m_thumbBarButtons which automatically
deletes the stored buttons in its destructor.

Closes https://github.com/wxWidgets/wxWidgets/pull/778

vadz commented:
How can the original warning be reproduced, exactly? Just configuring with `--enable-stl` and building `xmldll_xml.o` doesn't give any warnings for me, are any other configure options required or do I need to compile something else?

Also, Travis CI failures seem to be genuine, so this change seems to have broken something, did you look at this by chance?

vadz committed:
Canonicalize PBForDev email address in Git mailmap

Use the same name/email combination instead of several similar but
different ones.

rbresalier commented:
In my program, which I can't share (proprietary), I have instantiated a wxXmlDocument and used it. I also set wxUSE_STD_CONTAINERS as 1 in setup.h and built. My time is limited, maybe next weekend I can come up with a simple test case and look into the CI failures.

rbresalier commented:
I think it could also be reproduced outside of wxXmlDocument and textbuf if you instantiate one of these arrays and try to run the Sort function passing a CMPFUNC to it.

vadz commented:
If you could provide a minimal snippet of code reproducing the warning, it would be most helpful. TIA!

vadz committed:
Normalize email case of the commits done by PB

Canonicalize all the existing versions of the email used by PB to use
PBfordev as the local part.

rbresalier commented:
The warning can be reproduced by changing the warning level in Visual C compilation to /Wall and building the minimal sample with no source code change. Search the build output for "CMPFUNC'" and you will find many messages like these:

warning C4191: 'type cast': unsafe conversion from 'wxBaseArrayChar::CMPFUNC' to 'wxBaseArrayChar::SCMPFUNC'           
warning C4191: 'type cast': unsafe conversion from 'wxBaseArrayChar::CMPFUNC' to 'wxBaseArrayChar::SCMPFUNC'           
warning C4191: 'type cast': unsafe conversion from 'wxBaseArrayShort::CMPFUNC' to 'wxBaseArrayShort::SCMPFUNC'         
warning C4191: 'type cast': unsafe conversion from 'wxBaseArrayShort::CMPFUNC' to 'wxBaseArrayShort::SCMPFUNC'         
warning C4191: 'type cast': unsafe conversion from 'wxBaseArrayInt::CMPFUNC' to 'wxBaseArrayInt::SCMPFUNC'             
warning C4191: 'type cast': unsafe conversion from 'wxBaseArrayInt::CMPFUNC' to 'wxBaseArrayInt::SCMPFUNC'             
warning C4191: 'type cast': unsafe conversion from 'wxBaseArrayLong::CMPFUNC' to 'wxBaseArrayLong::SCMPFUNC'           

My proprietary program is only using /W3, and not /Wall, so I don't understand why the warning comes out with /W3 and my code but not with minimal that is using /W4. I tried comparing my compiler switches with the minimal sample and making the minimal sample's switches more like mine, but could not get it to reproduce unless I changed the warning level to /W4.

I think this is a real issue for the reasons that I described in my original message.

Perhaps this weekend I can try to understand why the test is failing and provide a fix.

rbresalier commented:
Figured out why program generates this error with /W3. I am #including some headers from cryptopp library, which enables some "security" warnings:

	// Security related, possible defects
	// http://blogs.msdn.com/b/vcblog/archive/2010/12/14/off-by-default-compiler-warnings-in-visual-c.aspx
#	pragma warning(once: 4191 4242 4263 4264 4266 4302 4826 4905 4906 4928)

This enables warning 4191 and that is why I see the warning in my code.

Based on the link provided in the above comment (copied below), it may make sense for wx to turn on these warnings as well. It would have caught this issue much earlier:

http://blogs.msdn.com/b/vcblog/archive/2010/12/14/off-by-default-compiler-warnings-in-visual-c.aspx

rbresalier committed:
_WX_DEFINE_SORTED_TYPEARRAY_2 was improperly performing casting "(CMPFUNC)m_fnCompare"
in 3 functions, converting a function with parameters of (T item1, T item2)
to (T *item1, T *item2). This was causing some array tests to fail after fixing
the warnings in previous commits. This commit makes changes to perform these
conversions properly.

rbresalier committed:
Merge branch 'fixCMPFUNCwarning' of github.com:rbresalier/wxWidgets into fixCMPFUNCwarning

Conflicts:
	include/wx/dynarray.h

vadz commented:
I think the original problem was fixed in 93edcaef2049c3f01ce0941ee1954e7253df74f3, please reopen or create another PR if you still see any problems and/or think anything here still should be applied. Thanks!
